### PR TITLE
Week 1

### PR DESCRIPTION
**Pull request by LeoKuD**
[Link to deploy](https://tic-tac-toe-leokud.herokuapp.com/)
Done: 
- play with computer;
- play with friend;
- win counter;
- history navigation;
- reset;
- save to localStorage;
- two levels of bot difficulty;
- select mark.

### CODE REVIEW

File: src/components/boadr/index.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        </div>
+        <div className="board-row">
+          {renderSquare(3)}
+          {renderSquare(4)}
+          {renderSquare(5)}
+        </div>
+        <div className="board-row">
+          {renderSquare(6)}
+          {renderSquare(7)}
+          {renderSquare(8)}
```
Comment (maximsan) [inline src/components/boadr/index.jsx:25]: you can just use array.map(...) to generate all items almost in one row of code

File: src/components/navigation/index.jsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    <div className={style.navigation}>
+      <div className={style.mode}>
+        <button onClick={modeHandlerClick}>
+          Mode: {isComputerMode ? MODE.computer : MODE.players}
+        </button>
+      </div>
+      <button onClick={resetGameHandleClick}>'Reset'</button>
+      <button
+        className={stepNumber ? '' : style.disabled}
+        disabled={stepNumber ? false : true}
```
Comment (maximsan) [inline src/components/navigation/index.jsx:23]: you can do like this `disabled={!!stepNumber}`

File: src/components/navigation/index.jsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        onClick={jumpToBackHandleClick}
+      >
+        'Back'
+      </button>
+      <button
+        className={stepNumber + 1 < history.length ? '' : style.disabled}
+        disabled={stepNumber + 1 < history.length ? false : true}
```
Comment (maximsan) [inline src/components/navigation/index.jsx:30]: don't repeat `stepNumber + 1 < history.length` twice

File: src/constants.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+export const TICTACTOEKEYS = {
```
Comment (maximsan) [inline src/constants.js:1]: why so strange? use simpler name and add space `GAME_KEYS`

File: src/constants.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    x: 'X',
+    o: 'O',
+    tie: 'tie'
+}
+
+export const PLAYERSKEYS = {
```
Comment (maximsan) [inline src/constants.js:7]: why so strange?  add space `PLAYERS_KEYS`

File: src/helpers.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        Math.floor(Math.random() * computerWinnerСombinations.length)
+      ].filter((item) => !currentGameState[item]);
+      random = random[Math.floor(Math.random() * random.length)];
+      index = random;
+    } else if (playerWinnerСombinations.length) {
+      let random = playerWinnerСombinations[
+        Math.floor(Math.random() * playerWinnerСombinations.length)
+      ].filter((item) => !currentGameState[item]);
+      random = random[Math.floor(Math.random() * random.length)];
+      index = random;
```
Comment (maximsan) [inline src/helpers.js:70]: why do you need this extra line ? create helper method, you repeat the same calculations for different combinations

File: src/index.css [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+.status {
+  margin-bottom: 10px;
+}
+
+.square {
+  background: #000;
+  border: 5px solid #fff;
+  color: #fff;
```
Comment (maximsan) [inline src/index.css:33]: if you repeat the same value more then once in several places please consider using :root selector with variables;

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+          squares: Array(9).fill(null),
+        },
+      ],
+      nextPlayer: true,
+      stepNumber: 0,
+      isComputerMode: false,
+      isNavigationMode: false,
+      isStart: false,
+      playerOne: true,
+    };
```
Comment (maximsan) [inline src/index.js:31]: here you can also call the method I mentioned below

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+class Game extends React.Component {
+  constructor(props) {
+    super(props);
+    this.score = {
+      x: 0,
+      o: 0,
+      tie: 0,
+      roundCompleted: false,
+    };
```
Comment (maximsan) [inline src/index.js:17]: here you can also call the method to init a state I mentioned below

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    if (
+      this.state.isStart &&
+      this.state.isComputerMode &&
+      this.state.nextPlayer &&
+      !this.state.isNavigationMode
+    ) {
+      this.computerMove();
+    }
+
+    if (prevState !== this.state) {
```
Comment (maximsan) [inline src/index.js:75]: I'm pretty sure that they different all the time, because they are objects 😄

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    const current = history[history.length - 1];
+    const squares = current.squares.slice();
+
+    if (calculateWinner(squares) || squares[i]) {
+      return;
+    }
+
+    if (this.state.isStart) {
+      squares[i] = this.state.nextPlayer ? TICTACTOEKEYS.x : TICTACTOEKEYS.o;
+      this.setState({
```
Comment (maximsan) [inline src/index.js:91]: use callback version of setState to update state with previous state

File: src/index.js [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        ]),
+        stepNumber: history.length,
+        nextPlayer: !this.state.nextPlayer,
+        isNavigationMode: false,
+      });
+    }
+  }
+
+  getStart() {
+    this.setState({ isStart: !this.state.isStart });
```
Comment (maximsan) [inline src/index.js:105]: we don't call this.state in this.setState. it's `anti-pattern`. you should use callback version of setState to update state with previous state

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  changeLavel() {
+    this.isLevelHard = !this.isLevelHard;
```
Comment (maximsan) [inline src/index.js:109]: it should not affect you UI, you should toggle it only inside `this.setState` by using callback version of `setState`

File: src/index.js [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  changePlayerMark(mark) {
+    this.setState({ nextPlayer: mark, playerOne: mark });
+  }
+
+  changeMode() {
+    this.setState({
+      isComputerMode: !this.state.isComputerMode,
```
Comment (maximsan) [inline src/index.js:118]: you shouldn't relay on this.state, you should update state by using callback version of `setState`

File: src/index.js [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+      tie: 0,
+      roundCompleted: false,
+    };
+    localStorage.removeItem('score');
+    this.resetGame();
+  }
+
+  getRandomEasy() {
+    const history = this.state.history;
+    const current = history[this.state.stepNumber].squares;
```
Comment (maximsan) [inline src/index.js:132]: you shouldn't relay on this.state, you should update state by using `setState`

File: src/index.js [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+
+    return getIndexEasyLevel(current);
+  }
+
+  getRandomHard() {
+    const history = this.state.history;
+    const current = history[this.state.stepNumber].squares;
```
Comment (maximsan) [inline src/index.js:139]: you shouldn't relay on this.state, you should update state by using `setState`

File: src/index.js [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+    return getIndexHardLevel(current, this.state.stepNumber);
+  }
+
+  computerMove() {
+    const board = document.querySelector('.game-board');
```
Comment (maximsan) [inline src/index.js:145]: create object with keys for all your selectors

File: src/index.js [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+    board.classList.toggle('blocked');
```
Comment (maximsan) [inline src/index.js:146]: create object with keys for all your specific classes

File: src/index.js [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+      this.handleClick(next);
+    }, 500);
+  }
+
+  resetGame() {
+    this.setState({
+      history: [{ squares: Array(9).fill(null) }],
+      stepNumber: 0,
+      nextPlayer: this.state.playerOne,
+      isNavigationMode: false,
```
Comment (maximsan) [inline src/index.js:159]: again use the init state method

File: src/index.js [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+    });
+    this.score.roundCompleted = false;
+  }
+
+  jumpTo(direction) {
+    let currentMove = this.state.stepNumber;
+    this.setState({
+      stepNumber: direction === 'back' ? --currentMove : ++currentMove,
+      nextPlayer: this.state.playerOne ? currentMove % 2 === 0 : !currentMove % 2 === 0,
+      isNavigationMode: true,
```
Comment (maximsan) [inline src/index.js:169]: you shouldn't relay on this.state, you should update state by using callback version of `setState`

File: src/index.js [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+      history: [{ squares: Array(9).fill(null) }],
+      stepNumber: 0,
+      nextPlayer: this.state.playerOne,
+      isNavigationMode: false,
+    });
+    this.score.roundCompleted = false;
+  }
+
+  jumpTo(direction) {
+    let currentMove = this.state.stepNumber;
```
Comment (maximsan) [inline src/index.js:165]: do destructuring assignment here

File: src/index.js [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+    let status;
+
+    if (winner && !this.score.roundCompleted) {
+      status = `Winner: ${winner}`;
+      this.updateScore(winner);
+    } else {
+      status = `Next player: ${
+        this.state.nextPlayer ? TICTACTOEKEYS.x : TICTACTOEKEYS.o
+      }`;
+    }
```
Comment (maximsan) [inline src/index.js:205]: create a method for this outside render

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some comment! Please review and fix ASAP. Thanks! UI looks dark and it's  🆒

File: src/components/boadr/index.jsx (hunk 1 above)
Comment (LeoKuD) [reply src/components/boadr/index.jsx:25]:
```suggestion
          export function Board({ onClick, squares }) {
  return (
    <div className={style.board}>
      {squares.map((item, index) => {
        function clickHandler() {
          onClick(index);
        }
        return <Square key={index} value={item} onClick={clickHandler} />;
      })}
    </div>
  );
}

```

File: src/components/navigation/index.jsx (hunk 2 above)
Comment (LeoKuD) [reply src/components/navigation/index.jsx:23]:
```suggestion
                disabled={!stepNumber}
```

File: src/components/navigation/index.jsx (hunk 3 above)
Comment (LeoKuD) [reply src/components/navigation/index.jsx:30]:
```suggestion
        className={isNextStep ? '' : style.disabled}
        disabled={!isNextStep}
```

File: src/constants.js (hunk 4 above)
Comment (LeoKuD) [reply src/constants.js:1]:
```suggestion
export const GAME_KEYS = {
```

File: src/constants.js (hunk 5 above)
Comment (LeoKuD) [reply src/constants.js:7]:
```suggestion
export const PLAYERS_KEYS = {
```

File: src/helpers.js [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+        availableСombinationsO.push([a, b, c]);
+      }
+    }
+
+    if (computerWinnerСombinations.length) {
+      let random = computerWinnerСombinations[
+        Math.floor(Math.random() * computerWinnerСombinations.length)
+      ].filter((item) => !currentGameState[item]);
+      random = random[Math.floor(Math.random() * random.length)];
+      index = random;
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/helpers.js:64]:
```suggestion
          index = randomX[Math.floor(Math.random() * randomX.length)];
```

File: src/helpers.js (hunk 6 above)
Comment (LeoKuD) [reply src/helpers.js:70]:
```suggestion
      return getRandomIndex(computerWinnerСombinations, currentGameState)
```

Comment (LeoKuD) [reply src/helpers.js:70]: function getRandomIndex(combinations, currentGameState) {
    let randomLine = combinations[
      Math.floor(Math.random() * combinations.length)
    ].filter((item) => !currentGameState[item]);

    return randomLine[Math.floor(Math.random() * randomLine.length)];
  }

File: src/helpers.js [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+
+  export function getIndexEasyLevel(currentGameState) {
+    let emptySquares = [];
+    currentGameState.forEach((element, index) => {
+      if (!element) {
+        emptySquares.push(index);
+      }
+    });
+    const randomSquares =
+      emptySquares[Math.floor(Math.random() * emptySquares.length)];
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/helpers.js:99]:
```suggestion
      return calculateRandomIndex(emptySquares);
```

File: src/index.css (hunk 7 above)
Comment (LeoKuD) [reply src/index.css:33]: Fixed. Created :root selector in the root css file and used it in other files.

```
:root {
  --bg-color-primary: #000;
  --color-primary: #fff;
}
```

File: src/index.js [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+        history: [
+          {
+            squares: Array(9).fill(null),
+          },
+        ],
+        nextPlayer: true,
+        stepNumber: 0,
+        isComputerMode: false,
+        isStart: false,
+        playerOne: true,
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:60]:
```
  initialState() {
    this.score = JSON.parse(localStorage.getItem('score')) || {
      x: 0,
      o: 0,
      tie: 0,
      roundCompleted: false,
    };
    this.isLevelHard = true;
    
    return JSON.parse(localStorage.getItem('state')) || { 
      history: [
        {
          squares: Array(9).fill(null),
        },
      ],
      nextPlayer: true,
      stepNumber: 0,
      isComputerMode: false,
      isNavigationMode: false,
      isStart: false,
      playerOne: true,
    };
  }
```

`    this.state = this.initialState();`

[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:60]: And deleted componentDidMount()

File: src/index.js [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+    this.setPlayerMarkOHandleClick = this.changePlayerMark.bind(this, false);
+    this.getStartHandleClick = this.getStart.bind(this);
+    this.changeLavelHandleClick = this.changeLavel.bind(this);
+  }
+
+  componentDidMount() {
+    this.score = JSON.parse(localStorage.getItem('score')) || {
+      x: 0,
+      o: 0,
+      tie: 0,
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:47]: corrected above

File: src/index.js (hunk 8 above)
Comment (LeoKuD) [reply src/index.js:31]: corrected above

File: src/index.js (hunk 9 above)
Comment (LeoKuD) [reply src/index.js:17]: corrected above

File: src/index.js (hunk 10 above)
Comment (LeoKuD) [reply src/index.js:75]: Yes, deleted this if.

Comment (LeoKuD) [reply src/index.js:75]:
```suggestion
        localStorage.setItem('state', JSON.stringify(this.state));
```

File: src/index.js (hunk 11 above)
Comment (LeoKuD) [reply src/index.js:91]:
```suggestion
           this.setState((state) => {
        if (state.isStart) {
          squares[i] = this.state.nextPlayer ? GAME_KEYS.x : GAME_KEYS.o;
          return {
            history: history.concat([
              {
                squares: squares,
              },
            ]),
            stepNumber: history.length,
            nextPlayer: !state.nextPlayer,
            isNavigationMode: false,
          };
        }
      });
```

File: src/index.js (hunk 12 above)
Comment (LeoKuD) [reply src/index.js:105]:
```
    this.setState((state) => {
      if (!state.isStart) {
        this.resetGame();
      }
      return { isStart: !state.isStart };
    });
```

File: src/index.js (hunk 13 above)
Comment (LeoKuD) [reply src/index.js:109]:
```suggestion
      changeLavel() {
    this.setState((state) => {
      return { isLevelHard: !state.isLevelHard };
    });
  }
```

File: src/index.js (hunk 14 above)
Comment (LeoKuD) [reply src/index.js:118]:
```suggestion
        changeMode() {
    this.setState((state) => {
      return {
        isComputerMode: !state.isComputerMode,
      }
    });
```

File: src/index.js (hunk 15 above)
Comment (LeoKuD) [reply src/index.js:132]:
```suggestion
      getRandomEasy() {
    let current = null;
    this.setState((state) => {
      current = state.history[state.stepNumber].squares;
    })
    return getIndexEasyLevel(current);
  }
```

File: src/index.js (hunk 16 above)
Comment (LeoKuD) [reply src/index.js:139]:
```suggestion
      getRandomHard() {
    let current = null;
    let stepNumber = null;
    this.setState((state) => {
      stepNumber = state.stepNumber;
      current = state.history[state.stepNumber].squares;
    })

    return getIndexHardLevel(current, stepNumber);
  }
```

File: src/index.js (hunk 17 above)
Comment (LeoKuD) [reply src/index.js:145]:
```
  export const SELECTOR_KEYS = {
    game: 'game',
    status: 'status',
    gameBoard: 'game-board',
    blocked: 'blocked'
  }
```

Comment (LeoKuD) [reply src/index.js:145]:
```
export const SELECTORS_KEYS = {
    game: 'game',
    status: 'status',
    gameBoard: 'game-board',
    blocked: 'blocked'
  }
```

File: src/index.js (hunk 18 above)
Comment (LeoKuD) [reply src/index.js:146]: corrected above

File: src/index.js (hunk 19 above)
Comment (LeoKuD) [reply src/index.js:159]: I can't use it here because method initialState rewrite all state

File: src/index.js (hunk 20 above)
Comment (LeoKuD) [reply src/index.js:169]:
```suggestion
        jumpTo(direction) {
    this.setState((state) => {
      let currentMove = state.stepNumber;
      return {
        stepNumber: direction === NAVIGATION_KEYS.back ? --currentMove : ++currentMove,
        nextPlayer: state.playerOne
          ? currentMove % 2 === 0
          : !currentMove % 2 === 0,
        isNavigationMode: true,
      };
    });
  }
```

File: src/index.js (hunk 21 above)
Comment (LeoKuD) [reply src/index.js:165]:
```suggestion
      let {stepNumber} = state;
```

File: src/index.js (hunk 22 above)
Comment (LeoKuD) [reply src/index.js:205]:
```
  updateStatus() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);
    let status;

    if (winner && !this.score.roundCompleted) {
      status = `Winner: ${winner}`;
      this.updateScore(winner);
    } else {
      status = `Next player: ${
        this.state.nextPlayer ? GAME_KEYS.x : GAME_KEYS.o
      }`;
    }

    return status
  }
```

File: src/index.js (hunk 19 above)
Comment (LeoKuD) [reply src/index.js:165]:
```
  resetGame() {
    this.setState((state) => {
      return {
        ...initialState(),
        nextPlayer: state.playerOne,
        isComputerMode: state.isComputerMode,
        isLevelHard: state.isLevelHard,
        isStart: state.isStart,
      };
    });
```

Comment (LeoKuD) [reply src/index.js]:
```
  resetGame() {
    this.setState((prevState) => {
      if (prevState.stepNumber) {
        return {
          ...initialState(),
          nextPlayer: prevState.playerOne,
          isComputerMode: prevState.isComputerMode,
          isLevelHard: prevState.isLevelHard,
          isStart: prevState.isStart,
        };
      }
    });
```

File: src/helpers.js [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+    return getIndexEasyLevel(currentGameState);
+  }
+}
+
+function getRandomIndex(combinations, currentGameState) {
+  let randomLine = combinations[
+    Math.floor(Math.random() * combinations.length)
+  ].filter((item) => !currentGameState[item]);
+
+  return calculateRandomIndex(randomLine);
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/helpers.js:128]: Fixed.
```suggestion
  function getRandomIndex(combinations, currentGameState) {
  let randomLine = calculateRandomIndex(combinations).filter(
    (item) => !currentGameState[item]
  );

  return calculateRandomIndex(randomLine);
}
```

File: src/helpers.js [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+      ? getRandomIndex(availableСombinationsX, currentGameState)
+      : getRandomIndex(availableСombinationsO, currentGameState);
+  } else {
+    return getIndexEasyLevel(currentGameState);
+  }
+}
+
+function getRandomIndex(combinations, currentGameState) {
+  let randomLine = combinations[
+    Math.floor(Math.random() * combinations.length)
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/helpers.js:125]:
```suggestion
    function getRandomIndex(combinations, currentGameState) {
  let randomLine = calculateRandomIndex(combinations).filter(
    (item) => !currentGameState[item]
  );

  return calculateRandomIndex(randomLine);
}
```

File: src/helpers.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/helpers.js:128]: Yes, sorry. Fixed.

File: src/helpers.js [hunk 29]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+export function getIndexEasyLevel(currentGameState) {
+  let emptySquares = [];
+
+  currentGameState.forEach((element, index) => {
+    if (!element) {
+      emptySquares.push(index);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/helpers.js:138]:
```suggestion
  export function getIndexEasyLevel(currentGameState) {
  let emptySquares = currentGameState
    .map((item, index) => (item ? null : index))
    .filter((item) => item !== null);
  return calculateRandomIndex(emptySquares);
}
```

File: src/index.js [hunk 30]
```diff
@@ -0,0 +1,260 @@
...
+    const current = history[history.length - 1];
+    const squares = current.squares.slice();
+
+    if (calculateWinner(squares) || squares[i]) {
+      return;
+    } else {
+      this.setState((prevState) => {
+        squares[i] = prevState.nextPlayer ? GAME_KEYS.x : GAME_KEYS.o;
+        if (prevState.isStart) {
+          return {
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:72]:
```suggestion
                  return {
            ...prevState,
```

File: src/index.js [hunk 31]
```diff
@@ -0,0 +1,260 @@
...
+          return { isStart: true };
+        }
+      });
+    }
+  }
+
+  getStart() {
+    this.setState((prevState) =>
+      !prevState.isStart
+        ? { isStart: !prevState.isStart }
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:93]:
```suggestion
              !prevState.isStart
        ? { ...prevState, isStart: !prevState.isStart }
        : {
            ...initialState(),
            nextPlayer: prevState.playerOne,
            isComputerMode: prevState.isComputerMode,
            isLevelHard: prevState.isLevelHard,
            isStart: !prevState.isStart,
          }
```

File: src/index.js [hunk 32]
```diff
@@ -0,0 +1,260 @@
...
+      ...initialStaticVariables(),
+    };
+    localStorage.removeItem(LOCALSTORAGE_KEYS.score);
+    this.resetGame();
+  }
+
+  getRandomEasy() {
+    let current = null;
+    this.setState((state) => {
+      current = state.history[state.stepNumber].squares;
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:128]: Yes, this is wrong. Deleted this function.

File: src/index.js [hunk 33]
```diff
@@ -0,0 +1,260 @@
...
+    this.setState((prevState) => {
+      if (prevState.stepNumber) {
+        return {
+          ...initialState(),
+          nextPlayer: prevState.playerOne,
+          isComputerMode: prevState.isComputerMode,
+          isLevelHard: prevState.isLevelHard,
+          isStart: prevState.isStart,
+        };
+      }
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:166]:
```suggestion
        resetGame() {
    this.setState((prevState) =>
      prevState.stepNumber
        ? {
            ...initialState(),
            nextPlayer: prevState.playerOne,
            isComputerMode: prevState.isComputerMode,
            isLevelHard: prevState.isLevelHard,
            isStart: prevState.isStart,
          }
        : {
            ...initialState(),
          }
    );
    this.score.roundCompleted = false;
  }
```

File: src/index.js [hunk 34]
```diff
@@ -0,0 +1,260 @@
...
+    });
+    this.score.roundCompleted = false;
+  }
+
+  jumpTo(direction) {
+    this.setState((state) => {
+      let { stepNumber } = state;
+      return {
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:174]:
```suggestion
            return {
        ...prevState,
```

File: src/index.js [hunk 35]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    return status;
+  }
+
+  render() {
+    const history = this.state.history;
+    const current = history[this.state.stepNumber];
+
+    let status = this.updateStatus();
```
[the remark this answers is not in the document]
Comment (LeoKuD) [reply src/index.js:226]: Deleted extra variable.
```suggestion
     render() {
    return (
```

**Review by maximsan — APPROVED** [review]
Looks good and It's approved!  You need 1 more approve 🙂 Let me know when you will have 2 approves

File: src/helpers.js [hunk 36]
```diff
@@ -0,0 +1,147 @@
...
+      return squares[a];
+    }
+  }
+  if (countSteps.length === 9) {
+    return 'tie';
+  }
+  return null;
+}
+
+export function getIndexHardLevel(currentGameState, stepNumber) {
```
Comment (Dev1LtJ) [inline src/helpers.js:45]: What does this function do?

File: src/helpers.js [hunk 37]
```diff
@@ -0,0 +1,147 @@
...
+      (currentGameState[b] === 'O' &&
+        !currentGameState[a] &&
+        !currentGameState[c]) ||
+      (currentGameState[c] === 'O' &&
+        !currentGameState[a] &&
+        !currentGameState[b])
+    ) {
+      availableСombinationsO.push([a, b, c]);
+    }
+  }
```
Comment (Dev1LtJ) [inline src/helpers.js:105]: You have a lot of repetitive code. Seems like you can simplify it. Try, please.

File: src/components/gameInfo/index.jsx [hunk 38]
```diff
@@ -0,0 +1,50 @@
...
+      <Settings
+        playerOne={playerOne}
+        setPlayerMarkXHandleClick={setPlayerMarkXHandleClick}
+        setPlayerMarkOHandleClick={setPlayerMarkOHandleClick}
+        history={history}
+        isStart={isStart}
+        getStartHandleClick={getStartHandleClick}
+        isComputerMode={isComputerMode}
+        changeLavelHardHandleClick={changeLavelHardHandleClick}
+        changeLavelEasyHandleClick={changeLavelEasyHandleClick}
```
Comment (Dev1LtJ) [inline src/components/gameInfo/index.jsx:45]: level?

File: src/index.js [hunk 39]
```diff
@@ -0,0 +1,246 @@
...
+            isComputerMode: prevState.isComputerMode,
+            isLevelHard: prevState.isLevelHard,
+            isStart: prevState.isStart,
+          }
+        : {
+            ...prevState,
+            isStart: false,
+          }
+    );
+    this.score.roundCompleted = false;
```
Comment (Dev1LtJ) [inline src/index.js:155]: Why don't you save it in the Store?

**Review by Dev1LtJ — CHANGES_REQUESTED** [review]
Hello! I left some comments. Try to fix them and ping me once it's done.

File: src/components/gameInfo/index.jsx (hunk 38 above)
Comment (LeoKuD) [reply src/components/gameInfo/index.jsx:45]:
```suggestion
          changeLevelHardHandleClick={this.changeLevelHardHandleClick}
          changeLevelEasyHandleClick={this.changeLevelEasyHandleClick}
```

File: src/index.js (hunk 39 above)
Comment (LeoKuD) [reply src/index.js:155]: Ok, moved to State.

```suggestion
      return { ...prevState, roundCompleted: true };
```

File: src/helpers.js (hunk 37 above)
Comment (LeoKuD) [reply src/helpers.js:105]:
```suggestion
     for (let i = 0; i < LINES.length; i++) {
    const [a, b, c] = LINES[i];
    if (checkPlayerWinnerCombitation(GAME_KEYS.o, [a, b, c])) {
      playerWinnerСombinations.push(
        checkPlayerWinnerCombitation(GAME_KEYS.o, [a, b, c])
      );
    }
    if (checkPlayerWinnerCombitation(GAME_KEYS.x, [a, b, c])) {
      computerWinnerСombinations.push(
        checkPlayerWinnerCombitation(GAME_KEYS.x, [a, b, c])
      );
    }
    if (checkAvailableСombinations(GAME_KEYS.x, [a, b, c])) {
      availableСombinationsX.push(
        checkAvailableСombinations(GAME_KEYS.x, [a, b, c])
      );
    }
    if (checkAvailableСombinations(GAME_KEYS.o, [a, b, c])) {
      availableСombinationsO.push(
        checkAvailableСombinations(GAME_KEYS.o, [a, b, c])
      );
    }
  }
```

**Review by oleg3555 — APPROVED** [review]
Hello, good work!)

**Pull request merged**